File: node-js-deposit-redeem-example/status.js
require("dotenv").config();
const { ethers } = require("ethers");

const rwaAbi = require("./abis/RWAToken.json").abi;
const whitelistAbi = require("./abis/Whitelist.json").abi;
const feeAbi = require("./abis/RWAFee.json").abi;

/**
 * Status script for pre-flight checks and previews.
 *
 * This is useful when integrators want to inspect everything before calling
 * deposit or redeem from their own backend or UI.
 */
function getEnv() {
  const env = {
    RPC_URL: process.env.RPC_URL,
    PRIVATE_KEY: process.env.PRIVATE_KEY,
    RWA_TOKEN_ADDRESS: process.env.RWA_TOKEN_ADDRESS,
    WHITELIST_ADDRESS: process.env.WHITELIST_ADDRESS,
    FEE_ADDRESS: process.env.FEE_ADDRESS || "",
    ENTRY_FEE_TIMING: Number(process.env.ENTRY_FEE_TIMING || 0),
    EXIT_FEE_TIMING: Number(process.env.EXIT_FEE_TIMING || 1),
  };

  if (!env.RPC_URL || !env.PRIVATE_KEY || !env.RWA_TOKEN_ADDRESS || !env.WHITELIST_ADDRESS) {
    throw new Error("Missing required env values. Please check .env.example");
  }

  return env;
}

async function resolveFeeAddress(vault, env) {
  if (env.FEE_ADDRESS) return env.FEE_ADDRESS;

  const feeAddress = await vault.rwaFee();
  if (!feeAddress || feeAddress === ethers.ZeroAddress) {
    throw new Error("Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address.");
  }

  return feeAddress;
}

async function main() {
  const desiredAssetsInput = process.argv[2] || "100";
  const sharesInput = process.argv[3] || "25";

  const env = getEnv();

  const provider = new ethers.JsonRpcProvider(env.RPC_URL);
  const wallet = new ethers.Wallet(env.PRIVATE_KEY, provider);

  const user = await wallet.getAddress();
  const vault = new ethers.Contract(env.RWA_TOKEN_ADDRESS, rwaAbi, wallet);
  const whitelist = new ethers.Contract(env.WHITELIST_ADDRESS, whitelistAbi, wallet);

  const assetAddress = await vault.asset();
  const asset = new ethers.Contract(
    assetAddress,
    [
      "function decimals() view returns (uint8)",
      "function symbol() view returns (string)",
      "function balanceOf(address) view returns (uint256)",
      "function allowance(address,address) view returns (uint256)"
    ],
    wallet
  );

  const shareDecimals = await vault.decimals();
  const shareSymbol = await vault.symbol();
  const assetDecimals = await asset.decimals();
  const assetSymbol = await asset.symbol();

  const desiredAssets = ethers.parseUnits(desiredAssetsInput, assetDecimals);
  const shares = ethers.parseUnits(sharesInput, shareDecimals);

  const feeAddress = await resolveFeeAddress(vault, env);
  const feeContract = new ethers.Contract(feeAddress, feeAbi, wallet);

  const [
    whitelisted,
    assetBalance,
    allowance,
    shareBalance,
    maxRedeem,
    entryFee,
    previewDepositShares,
    grossAssets,
    estimatedExitFee,
    previewRedeemNetAssets,
  ] = await Promise.all([
    whitelist.balanceOf(user, 1), // Token ID 1 is the relevant whitelist token
    asset.balanceOf(user),
    asset.allowance(user, env.RWA_TOKEN_ADDRESS),
    vault.balanceOf(user),
    vault.maxRedeem(user),
    feeContract.feeOnRaw(desiredAssets, env.ENTRY_FEE_TIMING),
    // NOTE: previewDeposit expects the total amount that will be sent to deposit()
    vault.previewDeposit(desiredAssets + (await feeContract.feeOnRaw(desiredAssets, env.ENTRY_FEE_TIMING))),
    vault.convertToAssets(shares),
    feeContract.feeOnTotal(await vault.convertToAssets(shares), env.EXIT_FEE_TIMING),
    vault.previewRedeem(shares),
  ]);

  const totalSpend = desiredAssets + entryFee;

  console.log("=== STATUS / DEBUG FLOW ===");
  console.log("User:", user);
  console.log("Whitelisted:", whitelisted);
  console.log("Vault:", env.RWA_TOKEN_ADDRESS);
  console.log("Whitelist:", env.WHITELIST_ADDRESS);
  console.log("Fee contract:", feeAddress);
  console.log("Asset token:", assetSymbol, assetAddress);
  console.log("Share token:", shareSymbol, env.RWA_TOKEN_ADDRESS);
  console.log("");
  console.log("--- Balances ---");
  console.log("Asset balance:", ethers.formatUnits(assetBalance, assetDecimals), assetSymbol);
  console.log("Allowance to vault:", ethers.formatUnits(allowance, assetDecimals), assetSymbol);
  console.log("Share balance:", ethers.formatUnits(shareBalance, shareDecimals), shareSymbol);
  console.log("maxRedeem:", ethers.formatUnits(maxRedeem, shareDecimals), shareSymbol);
  console.log("");
  console.log("--- Deposit preview ---");
  console.log("Desired assets before fee:", desiredAssetsInput, assetSymbol);
  console.log("Entry fee (feeOnRaw):", ethers.formatUnits(entryFee, assetDecimals), assetSymbol);
  console.log("Total spend:", ethers.formatUnits(totalSpend, assetDecimals), assetSymbol);
  console.log("previewDeposit(totalSpend):", ethers.formatUnits(previewDepositShares, shareDecimals), shareSymbol);
  console.log("");
  console.log("--- Redeem preview ---");
  console.log("Input shares:", sharesInput, shareSymbol);
  console.log("convertToAssets(shares) [gross info]:", ethers.formatUnits(grossAssets, assetDecimals), assetSymbol);
  console.log("Estimated exit fee (feeOnTotal):", ethers.formatUnits(estimatedExitFee, assetDecimals), assetSymbol);
  console.log("previewRedeem(shares) [final net]:", ethers.formatUnits(previewRedeemNetAssets, assetDecimals), assetSymbol);
}

main().catch((error) => {
  console.error("Status flow failed:", error.shortMessage || error.message || error);
  process.exit(1);
});

File: examples/python-deposit-redeem-example/status.py
import os
import sys
import json
from dotenv import load_dotenv
from web3 import Web3

load_dotenv()


def get_env():
    rwa_token = os.getenv("RWA_TOKEN_ADDRESS")
    whitelist = os.getenv("WHITELIST_ADDRESS")
    fee_address = os.getenv("FEE_ADDRESS") or ""

    env = {
        "RPC_URL": os.getenv("RPC_URL"),
        "PRIVATE_KEY": os.getenv("PRIVATE_KEY"),
        "RWA_TOKEN_ADDRESS": Web3.to_checksum_address(rwa_token) if rwa_token else "",
        "WHITELIST_ADDRESS": Web3.to_checksum_address(whitelist) if whitelist else "",
        "FEE_ADDRESS": Web3.to_checksum_address(fee_address) if fee_address else "",
        "ENTRY_FEE_TIMING": int(os.getenv("ENTRY_FEE_TIMING") or 0),
        "EXIT_FEE_TIMING": int(os.getenv("EXIT_FEE_TIMING") or 1),
    }

    if not all(
        [
            env["RPC_URL"],
            env["PRIVATE_KEY"],
            env["RWA_TOKEN_ADDRESS"],
            env["WHITELIST_ADDRESS"],
        ]
    ):
        raise ValueError("Missing required env values. Please check .env.example")

    return env


def load_abi(filename):
    path = os.path.join(os.path.dirname(__file__), "abis", filename)
    with open(path, "r") as f:
        data = json.load(f)
        return data.get("abi", data)


def resolve_fee_address(vault, env):
    if env["FEE_ADDRESS"]:
        return env["FEE_ADDRESS"]

    fee_address = vault.functions.rwaFee().call()
    if not fee_address or fee_address == "0x0000000000000000000000000000000000000000":
        raise ValueError(
            "Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address."
        )

    return fee_address


def main():
    desired_assets_input = sys.argv[2] if len(sys.argv) > 2 else "100"
    shares_input = sys.argv[3] if len(sys.argv) > 3 else "25"

    env = get_env()

    w3 = Web3(Web3.HTTPProvider(env["RPC_URL"]))
    account = w3.eth.account.from_key(env["PRIVATE_KEY"])
    user = account.address

    rwa_abi = load_abi("RWAToken.json")
    whitelist_abi = load_abi("Whitelist.json")
    fee_abi = load_abi("RWAFee.json")

    vault = w3.eth.contract(address=env["RWA_TOKEN_ADDRESS"], abi=rwa_abi)
    whitelist = w3.eth.contract(address=env["WHITELIST_ADDRESS"], abi=whitelist_abi)

    asset_address = vault.functions.asset().call()

    asset = w3.eth.contract(
        address=asset_address,
        abi=[
            {
                "inputs": [],
                "name": "decimals",
                "outputs": [{"type": "uint8"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [],
                "name": "symbol",
                "outputs": [{"type": "string"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [{"type": "address"}],
                "name": "balanceOf",
                "outputs": [{"type": "uint256"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [{"type": "address"}, {"type": "address"}],
                "name": "allowance",
                "outputs": [{"type": "uint256"}],
                "stateMutability": "view",
                "type": "function",
            },
        ],
    )

    share_decimals = vault.functions.decimals().call()
    share_symbol = vault.functions.symbol().call()
    asset_decimals = asset.functions.decimals().call()
    asset_symbol = asset.functions.symbol().call()

    desired_assets = int(float(desired_assets_input) * (10**asset_decimals))
    shares = int(float(shares_input) * (10**share_decimals))

    fee_address = resolve_fee_address(vault, env)
    fee_contract = w3.eth.contract(address=fee_address, abi=fee_abi)

    entry_fee = fee_contract.functions.feeOnRaw(
        desired_assets, env["ENTRY_FEE_TIMING"]
    ).call()
    total_spend = desired_assets + entry_fee

    preview_deposit_shares = vault.functions.previewDeposit(total_spend).call()

    gross_assets = vault.functions.convertToAssets(shares).call()

    estimated_exit_fee = fee_contract.functions.feeOnTotal(
        gross_assets, env["EXIT_FEE_TIMING"]
    ).call()

    preview_redeem_net_assets = vault.functions.previewRedeem(shares).call()

    whitelisted = whitelist.functions.balanceOf(user, 1).call()
    asset_balance = asset.functions.balanceOf(user).call()
    allowance = asset.functions.allowance(user, env["RWA_TOKEN_ADDRESS"]).call()
    share_balance = vault.functions.balanceOf(user).call()
    max_redeem = vault.functions.maxRedeem(user).call()

    print("=== STATUS / DEBUG FLOW ===")
    print("User:", user)
    print("Whitelisted:", whitelisted)
    print("Vault:", env["RWA_TOKEN_ADDRESS"])
    print("Whitelist:", env["WHITELIST_ADDRESS"])
    print("Fee contract:", fee_address)
    print("Asset token:", asset_symbol, asset_address)
    print("Share token:", share_symbol, env["RWA_TOKEN_ADDRESS"])
    print("")
    print("--- Balances ---")
    print("Asset balance:", asset_balance / (10**asset_decimals), asset_symbol)
    print("Allowance to vault:", allowance / (10**asset_decimals), asset_symbol)
    print("Share balance:", share_balance / (10**share_decimals), share_symbol)
    print("maxRedeem:", max_redeem / (10**share_decimals), share_symbol)
    print("")
    print("--- Deposit preview ---")
    print("Desired assets before fee:", desired_assets_input, asset_symbol)
    print("Entry fee (feeOnRaw):", entry_fee / (10**asset_decimals), asset_symbol)
    print("Total spend:", total_spend / (10**asset_decimals), asset_symbol)
    print(
        "previewDeposit(totalSpend):",
        preview_deposit_shares / (10**share_decimals),
        share_symbol,
    )
    print("")
    print("--- Redeem preview ---")
    print("Input shares:", shares_input, share_symbol)
    print(
        "convertToAssets(shares) [gross info]:",
        gross_assets / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "Estimated exit fee (feeOnTotal):",
        estimated_exit_fee / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "previewRedeem(shares) [final net]:",
        preview_redeem_net_assets / (10**asset_decimals),
        asset_symbol,
    )


if __name__ == "__main__":
    main()
